doc/sphinx/practical-tools/utilities.rst

```diff
@@ -7,32 +7,294 @@
+
+  Coq tools are limited to using only one `_CoqProject` at a time, so multi-package
+  projects need another mechanism, such as:
+
+- The COQPATH environment variable, which contains a list of directories.
+  Each of their subdirectories is a separately-compiled package.  The entire
+  logical path is derived from the directory name (as for `Bignums` above).
+  See :ref:`customization-by-environment-variables`.
+
+  TODO improve what's written at the hyperlink
+
+- :cmd:`Add LoadPath` adds a new logical path/filesystem path pair.  We recommend
+  you avoid using :cmd:`Add LoadPath` because embedding filesystem paths in your scripts
+  makes it harder to install and use them on another computer.
+
+We recommend using COQPATH.  For simpler developments, it could have a single
+entry pointing to a directory that contains each of your separate projects.
+Make sure you use distinct logical names for each project.  For more advanced
+development work, you may want to set COQPATH differently for different projects
+or different versions of a project.  Currently, if you want to change COQPATH
+while you're using CoqIDE, you must restart CoqIDE or start another CoqIDE process
+after making the change.  Perhaps in the future CoqIDE will show the current
+setting for and allow changing COQPATH.
+
+Alternatively, if you have a single-package project that depends on external
+packages that were not installed with the Coq binaries or accessible through
+COQPATH, you could add `-R` lines to the `_CoqProject` that point to the
+dependencies.  Of course, this makes you `_CoqProject` system dependent and
+those lines shouldn't appear in published packages.
+
+TODO I thought @herbelin added code to complain about ambiguous short names
+I made up some stuff below, need to check it:
+
+If you associate the same logical name with more than one directory, Coq will
+use the last one added (i.e., that appears earlier in :cmd:`Print LoadPath`).
+Nor will Coq complain if the short name of a package is ambiguous.  Perhaps
+in the future, Coq will be able to report such conflicts.  We recommend using
+the :n:`From @dirpath` option of :cmd:`Require` to make the behavior more predictable.
+
+
+Upgrading to a new version of Coq
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`.vo` files are specific to the version of Coq that compiled it.  When you
+upgrade to a new version of Coq, you must recompile all the projects
+that you want to run in the new version.  This is necessary to assure that
+your proofs still work in the new version.  Once their projects build on the
+new version, most users no longer have a need to run on the old version.
+
+If, however, you want to overlap working on your project on both the old and new
+versions, you'll need to create separate source directories for your project
+for the different Coq versions.  Currently the compiled `.vo` files are kept
+in the same directory as their corresponding `.v` file and the `.vo` is specific
+to one version of Coq.
```

Comment:
There are two ways to work with multiple versions: you can separate them by space or time.   "Space" meaning by the use of multiple copies in your file system, and "time" meaning you have a single copy that you update to the particular version you need at the moment.  Now that you mention it, IIRC one a switch is created, opam doesn't need to recompile it, right?  That would work neatly for any Coq package that was installed with opam.

But how about the code for a project that someone is actively developing?  You can make that a local opam repository, add that as an additional repository, configure the switch and then actually switch.    (Do they have to have it in git?  What if it's not in github?)  If you know of an existing explanation of how to do this or you can provide a short but reasonably complete explanation of how to do that, I can try to work that in.  Or I can just suggest the possibility without much detail.

Reply:
You don't have to make your own projects into opam projects to use Coq and it's packages installed in an opam switch. They will look like they were installed globally as usual. As long as your path variable is configured to point to the opam bin and lib folders, everything should work as with a usual install.

The pointing to lib and bin is exactly what the command `eval $(opam env)` does. Although I don't know what the windows equivalent for this would be. It should definitely work in cygwin AFAIK however.